File: WebDriverTests_Java/WebDriverTests/src/test/java/core/systemControls/HtmlControl.java
package core.systemControls;

import core.webDriver.WebDriverCache;
import core.webDriver.WebElementWrapper;
import core.supportClass.RefObject;
import org.openqa.selenium.By;
import org.openqa.selenium.SearchContext;
import org.openqa.selenium.WebElement;
import org.testng.Assert;
import ru.yandex.qatools.allure.annotations.Step;

public abstract class HtmlControl {
    private String controlDescription;
    public WebElementWrapper element;
    public By locator;
    public SearchContext searchContext;


    protected HtmlControl(By locator) {
        this(locator, null);
    }


    protected HtmlControl(By locator, HtmlControl container) {
        setContainer(container);
        setLocator(locator);
        Container = getContainer();
        this.locator = getLocator();
        controlDescription = FormatControlDescription(locator.toString(), container);
        searchContext = getContainer().getSearchContext();
        element = new WebElementWrapper(searchContext, locator, controlDescription);
    }

    private HtmlControl Container;
    public final HtmlControl getContainer()
    {
        return Container;
    }
    private void setContainer(HtmlControl value)
    {
        Container = value;
    }

    private By Locator;
    public final By getLocator()
    {
        return Locator;
    }
    private void setLocator(By value)
    {
        Locator = value;
    }
    protected HtmlControl(String idLocator) {
        this(idLocator, null);
    }

    protected HtmlControl(String idLocator, HtmlControl container) {
        this(By.id(idLocator), container);
    }

    public final String FormatControlDescription(String locatorString, HtmlControl container)
    {
        String description = String.format("%1$s (%2$s)", this.getClass().getSimpleName(), locatorString);
        if (container == null)
        {
            return description;
        }
        return String.format("%1$s в контексте элемента %2$s", description, container.controlDescription);
    }
    public static SearchContext getSearchContext(HtmlControl container)
    {
        return container != null ? container.element : WebDriverCache.getWebDriver().GetSearchContext();
    }
    public static SearchContext getSearchContext()
    {
        return WebDriverCache.getWebDriver().GetSearchContext();
    }

    public final void switchToFrame()
    {
        WebDriverCache.getWebDriver().switchToFrame(locator);
    }

    public final void javascriptExecutorClick()
    {
        WebElement button = searchContext.findElement(locator);
        WebDriverCache.getWebDriver().javascriptExecutor("arguments[0].click();", button);
    }

    public final Object javascriptExecuteWebElement(String script)
    {
        WebElement webElement = searchContext.findElement(locator);
        return WebDriverCache.getWebDriver().javascriptExecutor(script, webElement);
    }

    public final Object javascriptExecute(String script)
    {
        return WebDriverCache.getWebDriver().ExecuteScript(script);
    }

    public final void switchToDefaultContent()
    {
        WebDriverCache.getWebDriver().switchToDefaultContent();
    }

    public boolean getIsEnabled()
    {
        return !hasClass("disabled");
    }

    public final boolean getIsPresent()
    {
        return (getContainer() == null || getContainer().getIsPresent()) && searchContext.findElements(locator).size() > 0;
    }

    public boolean getIsVisible()
    {
        return element.isDisplayed();
    }

    public final SearchContext searchContext()
    {
        return searchContext;
    }
    public final void mouseover()
    {
        element.mouseover();
    }

    public final void sendKeysToBody(String text)
    {
        element.sendKeysToBody(text);
    }

    public final void sendKeys(String text)
    {
        element.sendKeys(text);
    }

    public String getText()
    {
        return element.getText().trim();
    }

    public String getTagName()
    {
        return element.getTagName().trim();
    }

    public String getValue()
    {
        return element.getAttribute("value");
    }

    public void AssertNessesary() throws InterruptedException {
        WaitClassPresenceWithRetries("field__incorrect");
        WaitAttributeValue("title", "Поле должно быть заполнено");
    }

    public final boolean hasClass(String className)
    {
        String actualClasses = null;
        RefObject<String> tempRef_actualClasses = new RefObject<String>(actualClasses);
        boolean tempVar = hasClass(className, tempRef_actualClasses);
        actualClasses = tempRef_actualClasses.argValue;
        return tempVar;
    }

    private boolean hasClass(String className, RefObject<String> actualClasses) {
        actualClasses.argValue = getAttributeValue("class");
        String[] actualClassesArray = actualClasses.argValue.split(String.valueOf(new String[]{" ", "\r", "\n", "\t"}), -1);
        String[] expectedClassesArray = className.split(String.valueOf(new String[] {" ", "\r", "\n", "\t"}), -1);
return true;
//        return expectedClassesArray.All(actualClassesArray.contains);
    }

    public String getAttributeValue(String attributeName)
    {
        return element.getAttribute(attributeName);
    }

    public String formatWithLocator(String text)
    {
        return String.format("'%1$s' '%2$s'", text, controlDescription);
    }

    @Step("Ждем видимости")
    public final void WaitVisible()
    {
        String description = formatWithLocator("Ожидание видимости элемента");
        Assert.assertEquals(getIsVisible(), true, description);
    }
    @Step("Ждем видимости")
    public final void WaitVisibleWithRetries()
    {
        WaitVisibleWithRetries(null);
    }
    @Step("Ожидание видимости элемента")
    public final void WaitVisibleWithRetries(Integer timeout)
    {
        String description = formatWithLocator("Ожидание видимости элемента");
        Waiter.Wait(()-> getIsVisible(), description, timeout);
    }
    @Step("Ожидание не видимости элемента")
    public final void WaitInvisible()
    {
        String description = formatWithLocator("Ожидание невидимости элемента");
        Assert.assertTrue(getIsVisible(), description);
    }

    @Step("Ожидание не видимости элемента")
    public final void WaitInvisibleWithRetries()
    {
        WaitInvisibleWithRetries(null);
    }

    @Step("Ожидание не видимости элемента")
    public final void WaitInvisibleWithRetries(Integer timeout) {
        String description = formatWithLocator("Ожидание невидимости элемента");
        Waiter.Wait(() -> !getIsVisible(), description, timeout);
    }

    @Step("Ожидание присутствие элемента")
    public final void WaitPresence()
    {
        String description = formatWithLocator("Ожидание присутствия элемента");
        Assert.assertTrue(getIsPresent(), description);
    }

    @Step("Ожидание присутствие элемента")
    public final void WaitPresenceWithRetries()
    {
        WaitPresenceWithRetries(null);
    }

    @Step("Ожидание присутствие элемента")
    public final void WaitPresenceWithRetries(Integer timeout)
    {
        String description = formatWithLocator("Ожидание присутствия элемента");
        Waiter.Wait(() -> getIsPresent(), description, timeout);
    }

    @Step("Ожидание отсутствия элемента")
    public final void WaitAbsence()
    {
        String description = formatWithLocator("Ожидание отсутствия элемента");
        Assert.assertFalse(getIsPresent(), description);
    }

    @Step("Ожидание отсутствия элемента")
    public final void WaitAbsenceWithRetries()
    {
        WaitAbsenceWithRetries(null);
    }

    @Step("Ожидание отсутствия элемента")
    public final void WaitAbsenceWithRetries(Integer timeout)
    {
        String description = formatWithLocator("Ожидание отсутствия элемента");
        Waiter.Wait(() -> !getIsPresent(), description, timeout);
    }

    @Step("Ожидание доступности элемента")
    public final void WaitEnabled()
    {
        String description = formatWithLocator("Ожидание доступности элемента");
        Assert.assertTrue(getIsEnabled(), description);
    }

    @Step("Ожидание доступности элемента")
    public final void WaitEnabledWithRetries()
    {
        WaitEnabledWithRetries(null);
    }

    @Step("Ожидание доступности элемента")
    public final void WaitEnabledWithRetries(Integer timeout)
    {
        String description = formatWithLocator("Ожидание доступности элемента");
        Waiter.Wait(() -> getIsEnabled(), description, timeout);
    }

    @Step("Ожидание недоступности элемента")
    public final void WaitDisabled()
    {
        String description = formatWithLocator("Ожидание недоступности элемента");
        Assert.assertFalse(getIsEnabled(), description);
    }

    @Step("Ожидание недоступности элемента")
    public final void WaitDisabledWithRetries()
    {
        WaitDisabledWithRetries(null);
    }

    @Step("Ожидание недоступности элемента")
    public final void WaitDisabledWithRetries(Integer timeout)
    {
        String description = formatWithLocator("Ожидание недоступности элемента");
        Waiter.Wait(() -> !getIsEnabled(), description, timeout);
    }
    @Step("Ожидание текста {0} в элемента")
    public void WaitText(String expectedText)
    {
        String description = formatWithLocator(String.format("Ожидание появления текста '%1$s' в элементе", expectedText));
        Waiter.Wait(() -> getText().equals(expectedText), description);
    }

    @Step("Ожидание value {0} в элементе")
    public final void WaitValue(String value)
    {
        String description = formatWithLocator(String.format("Ожидание value '%1$s' в элементе", value));
        Waiter.Wait(() -> getAttributeValue("value").equals(value), description);
    }

    @Step("Ожидание текста {0} в элементе")
    public final void WaitTextStartsWith(String expectedText)
    {
        String description = formatWithLocator(String.format("Ожидание текста '%1$s' в начале текста элемента", expectedText));
        Waiter.Wait(() -> getText().startsWith(expectedText), description);
    }

    @Step("Ожидание текста {0} в внутри текста элемента")
    public final void WaitTextContains(String expectedText)
    {
        String description = formatWithLocator(String.format("Ожидание текста '%1$s' внутри текста элемента", expectedText));
        Waiter.Wait(() -> getText().contains(expectedText), description);
    }

    @Step("Ожидание текста {0} в внутри текста элемента")
    public final void WaitTextContainsWithRetries(String expectedText)
    {
        WaitTextContainsWithRetries(expectedText, null);
    }

    @Step("Ожидание текста {0} в внутри текста элемента")
    public final void WaitTextContainsWithRetries(String expectedText, Integer timeout)
    {
        String description = formatWithLocator(String.format("Ожидание текста '%1$s' внутри текста элемента", expectedText));
        Waiter.Wait(() -> getText().contains(expectedText), description, timeout);
    }

    @Step("Ожидание класса {0}  у элемента")
    public void WaitClassPresence(String className)
    {
        String description = formatWithLocator(String.format("Ожидание класса '%1$s' у элемента", className));
        String actualClasses = null;
        RefObject<String> tempRef_actualClasses = new RefObject<String>(actualClasses);
        Assert.assertTrue(hasClass(className, tempRef_actualClasses), String.format("'%1$s', актуальный класс: '%2$s'", description, actualClasses));
        actualClasses = tempRef_actualClasses.argValue;
    }

    @Step("Ожидание класса {0} у элемента")
    public void WaitClassPresenceWithRetries(String className)
    {
        WaitClassPresenceWithRetries(className, null);
    }

    @Step("Ожидание класса {0} у элемента")
    public void WaitClassPresenceWithRetries(String className, Integer timeout)
    {
        String description = formatWithLocator(String.format("Ожидание появления класса '%1$s' у элемента", className));
        Waiter.Wait(() -> hasClass(className), description, timeout);
    }

    @Step("Ожидание отсутствия класса {0} у элемента")
    public final void WaitClassAbsence(String className)
    {
        String description = formatWithLocator(String.format("Ожидание отсутствия класса '%1$s' у элемента", className));
        String actualClasses = null;
        RefObject<String> tempRef_actualClasses = new RefObject<String>(actualClasses);
        Assert.assertFalse(hasClass(className, tempRef_actualClasses), String.format("'%1$s', актуальный класс: '%2$s'", description, actualClasses));
        actualClasses = tempRef_actualClasses.argValue;
    }

    @Step("Ожидание отсутствия класса {0} у элемента")
    public final void WaitClassAbsenceWithRetries(String className)
    {
        WaitClassAbsenceWithRetries(className, null);
    }

    @Step("Ожидание отсутствия класса {0} у элемента")
    public final void WaitClassAbsenceWithRetries(String className, Integer timeout) {
        String description = formatWithLocator(String.format("Ожидание отсутствия класса '%1$s' у элемента", className));
        Waiter.Wait(() -> !hasClass(className), description, timeout);
    }

    @Step("Ожидание отсутствия класса {0} у элемента")
    public final void WaitAttributeValue(String attributeName, String expectedText)
    {
        String description = formatWithLocator(String.format("Ожидание аттрибута '%1$s' со значением '%2$s' в элементе", attributeName, expectedText));
        Assert.assertEquals(expectedText, getAttributeValue(attributeName), description);
    }

    @Step("Ожидание отсутствия класса {0} у элемента")
    public final void WaitAttributeValueWithRetries(String attributeName, String expectedText)
    {
        WaitAttributeValueWithRetries(attributeName, expectedText, null);
    }

    public final void WaitAttributeValueWithRetries(String attributeName, String expectedText, Integer timeout)
    {
        String description = formatWithLocator(String.format("Ожидание аттрибута '%1$s' со значением '%2$s' в элементе", attributeName, expectedText));
        Waiter.Wait(() -> expectedText == getAttributeValue(attributeName), description, timeout);
    }

    @Step("Кликнуть на элемент")
    public void click() {
        Waiter.WaitExeption(() -> element.click(), "Не удалось кликнуть на элемент");
    }

    @Step("Кликнуть на элемент")
    public void WaitIsClicked() {
        Waiter.WaitExeption(() -> element.click(), "Не удалось кликнуть на элемент");
    }
}

File: WebDriverTests_Csharp/Autotest.Core/Autotests.Utilities/WebTestCore/SystemControls/HtmlControl.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using Autotests.Utilities.WebTestCore.TestSystem;
using NUnit.Framework;
using OpenQA.Selenium;

namespace Autotests.Utilities.WebTestCore.SystemControls
{
    public abstract class HtmlControl
    {
        private readonly string controlDescription;
        public readonly WebElementWrapper element;
        private readonly By locator;
        private readonly ISearchContext searchContext;

        protected HtmlControl(By locator, HtmlControl container = null)
        {
            Container = container;
            this.locator = locator;
            controlDescription = FormatControlDescription(locator.ToString(), container);
            searchContext = container.GetSearchContext();
            element = new WebElementWrapper(searchContext, locator, controlDescription);
        }

        protected HtmlControl(string idLocator, HtmlControl container = null)
            : this(By.Id(idLocator), container)
        {
        }

        public void SwitchToFrame()
        {
            WebDriverCache.WebDriver.SwitchToFrame(SearchContext.FindElement(By.XPath("//*[@id='ddelivery-container']/iframe")));
        }

        public void SwitchToDefaultContent()
        {
            WebDriverCache.WebDriver.SwitchToDefaultContent();
        }

        public virtual bool IsEnabled
        {
            get { return !HasClass("disabled"); }
        }

        public bool IsPresent
        {
            get { return (Container == null || Container.IsPresent) && searchContext.FindElements(locator).Count > 0; }
        }

        public virtual bool IsVisible
        {
            get { return element.Displayed; }
        }

        public HtmlControl Container { get; private set; }

        public By Locator
        {
            get { return locator; }
        }

        public ISearchContext SearchContext
        {
            get { return searchContext; }
        }

        public void WaitVisible()
        {
            string description = FormatWithLocator("Ожидание видимости элемента");
            Assert.IsTrue(IsVisible, description);
        }

        public void WaitVisibleWithRetries(int? timeout = null)
        {
            string description = FormatWithLocator("Ожидание видимости элемента");
            Waiter.Wait(() => IsVisible, description, timeout);
        }

        public void WaitInvisible()
        {
            string description = FormatWithLocator("Ожидание невидимости элемента");
            Assert.IsFalse(IsVisible, description);
        }

        public void WaitInvisibleWithRetries(int? timeout = null)
        {
            string description = FormatWithLocator("Ожидание невидимости элемента");
            Waiter.Wait(() => !IsVisible, description, timeout);
        }

        public void WaitPresence()
        {
            string description = FormatWithLocator("Ожидание присутствия элемента");
            Assert.IsTrue(IsPresent, description);
        }

        public void WaitPresenceWithRetries(int? timeout = null)
        {
            string description = FormatWithLocator("Ожидание присутствия элемента");
            Waiter.Wait(() => IsPresent, description, timeout);
        }

        public bool CheckPresenceWithRetries()
        {
            bool result = true;
            Waiter.DoWait(() => IsPresent, () => { result = false; });
            return result;
        }

        public void WaitAbsence()
        {
            string description = FormatWithLocator("Ожидание отсутствия элемента");
            Assert.IsFalse(IsPresent, description);
        }

        public void WaitAbsenceWithRetries(int? timeout = null)
        {
            string description = FormatWithLocator("Ожидание отсутствия элемента");
            Waiter.Wait(() => !IsPresent, description, timeout);
        }

        public void WaitEnabled()
        {
            string description = FormatWithLocator("Ожидание доступности элемента");
            Assert.IsTrue(IsEnabled, description);
        }

        public void WaitEnabledWithRetries(int? timeout = null)
        {
            string description = FormatWithLocator("Ожидание доступности элемента");
            Waiter.Wait(() => IsEnabled, description, timeout);
        }

        public void WaitDisabled()
        {
            string description = FormatWithLocator("Ожидание недоступности элемента");
            Assert.IsFalse(IsEnabled, description);
        }

        public void WaitDisabledWithRetries(int? timeout = null)
        {
            string description = FormatWithLocator("Ожидание недоступности элемента");
            Waiter.Wait(() => !IsEnabled, description, timeout);
        }

        public virtual void WaitText(string expectedText)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание появления текста '{0}' в элементе", expectedText));
            Waiter.Wait(() => GetText() == expectedText, description);
        }

        public void WaitValue(string value)
        {
            string description = FormatWithLocator(string.Format("Ожидание value '{0}' в элементе", value));
            Waiter.Wait(() => GetAttributeValue("value") == value, description);
        }
        public void WaitTextStartsWith(string expectedText)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание текста '{0}' в начале текста элемента", expectedText));
            StringAssert.StartsWith(expectedText, GetText(), description);
        }

        public void WaitTextContains(string expectedText)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание текста '{0}' внутри текста элемента", expectedText));
            StringAssert.Contains(expectedText, GetText(), description);
        }

        public virtual void WaitTextNotContains(string expectedText)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание отсутствия текста '{0}' внутри текста элемента", expectedText));
            StringAssert.DoesNotContain(expectedText, GetText(), description);
        }

        public void WaitTextContainsWithRetries(string expectedText, int? timeout = null)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание текста '{0}' внутри текста элемента", expectedText));
            Waiter.Wait(() => GetText().Contains(expectedText), description, timeout);
        }

        public virtual void WaitClassPresence(string className)
        {
            string description = FormatWithLocator(string.Format("Ожидание класса '{0}' у элемента", className));
            string actualClasses;
            Assert.IsTrue(HasClass(className, out actualClasses),
                string.Format("{0}, актуальный класс: '{1}'", description, actualClasses));
        }

        public virtual void WaitClassPresenceWithRetries(string className, int? timeout = null)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание появления класса '{0}' у элемента", className));
            Waiter.Wait(() => HasClass(className), description, timeout);
        }

        public void WaitClassAbsence(string className)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание отсутствия класса '{0}' у элемента", className));
            string actualClasses;
            Assert.IsFalse(HasClass(className, out actualClasses),
                string.Format("{0}, актуальный класс: '{1}'", description, actualClasses));
        }

        public void WaitClassAbsenceWithRetries(string className, int? timeout = null)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание отсутствия класса '{0}' у элемента", className));
            Waiter.Wait(() => !HasClass(className), description, timeout);
        }

        public void WaitAttributeValue(string attributeName, string expectedText)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание аттрибута '{0}' со значением '{1}' в элементе", attributeName,
                    expectedText));
            Assert.AreEqual(expectedText, GetAttributeValue(attributeName), description);
        }

        public void WaitAttributeValueWithRetries(string attributeName, string expectedText, int? timeout = null)
        {
            string description =
                FormatWithLocator(string.Format("Ожидание аттрибута '{0}' со значением '{1}' в элементе", attributeName,
                    expectedText));
            Waiter.Wait(() => expectedText == GetAttributeValue(attributeName), description, timeout);
        }

        public virtual void Click()
        {
            element.Click();
        }

        public void ScrollDown()
        {
            element.ScrollDown();
        }

        public void ClickLeftUp()
        {
            element.ClickLeftUp();
        }

        public void ClickAndWaitTextError(int index = 0, int timeout = 6000, int waitTimeout = 100)
        {
            Click(); 
            var alertClass = new StaticControl(BY.NthOfClass("alert-error", index));
            var errorClass = new StaticControl(BY.NthOfClass("help-inline", index));
            var w = Stopwatch.StartNew();
            while (errorClass.IsPresent == false && alertClass.IsPresent == false)
            {
                Thread.Sleep(waitTimeout);
                if (w.ElapsedMilliseconds > timeout) Assert.AreEqual(errorClass.IsPresent, true, "Время ожидание завершено. Не найден элемент содержаший ошибку");
                if (w.ElapsedMilliseconds > timeout) Assert.AreEqual(alertClass.IsPresent, true, "Время ожидание завершено. Не найден элемент содержаший ошибку");
            }
        }

        public void ClickAndWaitTextHorizontalError(int index = 0, int timeout = 6000, int waitTimeout = 100)
        {
            Click(); 
            var alertClass = new StaticControl(BY.NthOfClass("alert-error", index));
            var w = Stopwatch.StartNew();
            while (alertClass.IsPresent == false)
            {
                Thread.Sleep(waitTimeout);
                if (w.ElapsedMilliseconds > timeout) Assert.AreEqual(alertClass.IsPresent, true, "Время ожидание завершено. Не найден элемент содержаший ошибку");
            }
        }

        public void ClickAndWaitTextErrorAbsence(int index = 0, int timeout = 6000, int waitTimeout = 100)
        {
            var alertClass = new StaticControl(BY.NthOfClass("alert-error", index));
            var errorClass = new StaticControl(BY.NthOfClass("help-inline", index));
            Click();
            var w = Stopwatch.StartNew();
            while (errorClass.IsPresent || alertClass.IsPresent)
            {
                Thread.Sleep(waitTimeout);
                if (w.ElapsedMilliseconds > timeout) Assert.AreEqual(errorClass.IsPresent, false, "Время ожидание завершено. Найден элемент содержаший ошибку");
                if (w.ElapsedMilliseconds > timeout) Assert.AreEqual(alertClass.IsPresent, false, "Время ожидание завершено. Найден элемент содержаший ошибку");
            }
        }

        public void Mouseover()
        {
            element.Mouseover();
        }

        public void SendKeysToBody(string text)
        {
            element.SendKeysToBody(text);
        }

        public void  SendKeys(string text)
        {
            element.SendKeys(text);
        }

        public virtual string GetText()
        {
            return element.Text.Trim();
        }
        
        public virtual string GetValue()
        {
            return element.GetAttribute("value");
        }
        
        public string GetAttributeValue(string attributeName)
        {
            return element.GetAttribute(attributeName);
        }

        public virtual void AssertNessesary()
        {
            WaitClassPresenceWithRetries("field__incorrect");
            WaitAttributeValue("title", "Поле должно быть заполнено");
        }

        public bool HasClass(string className)
        {
            string actualClasses;
            return HasClass(className, out actualClasses);
        }

        public string FormatWithLocator(string text)
        {
            return string.Format("{0} '{1}'", text, controlDescription);
        }

        public string GetParameterValue(string value)
        {
            string[] path = value.Split('.');
            string script = string.Format("return Configs['{0}']", GetAttributeValue("id"));
            script = path.Aggregate(script, (current, part) => current + string.Format("['{0}']", part));
            return WebDriverCache.WebDriver.ExecuteScript(script) as string;
        }

        private bool HasClass(string className, out string actualClasses)
        {
            actualClasses = GetAttributeValue("class");
            string[] actualClassesArray = actualClasses.Split(new[] {" ", "\r", "\n", "\t"},
                StringSplitOptions.RemoveEmptyEntries);
            string[] expectedClassesArray = className.Split(new[] {" ", "\r", "\n", "\t"},
                StringSplitOptions.RemoveEmptyEntries);
            return expectedClassesArray.All(actualClassesArray.Contains);
        }

        public string FormatControlDescription(string locatorString, HtmlControl container)
        {
            string description = string.Format("{0} ({1})", GetType().Name, locatorString);
            if (container == null)
                return description;
            return string.Format("{0} в контексте элемента {1}", description, container.controlDescription);
        }
    }
}